lib/xandra/connection/utils.ex

```diff
@@ -49,11 +49,31 @@ defmodule Xandra.Connection.Utils do
     # receive the response to this frame because if we said we want to use
     # compression, this response is already compressed.
     with :ok <- :gen_tcp.send(socket, payload),
-         {:ok, %Frame{body: <<>>}} <- recv_frame(socket, compressor) do
-       :ok
+         {:ok, frame} <- recv_frame(socket, compressor) do
+      case frame do
+        %Frame{body: <<>>} ->
+          :ok
+        %Frame{kind: :authenticate} ->
+          authenticate_connection(socket, requested_options, compressor, options)
+      end
     else
       {:error, reason} ->
         {:error, ConnectionError.new("startup connection", reason)}
     end
   end
+
+  defp authenticate_connection(socket, requested_options, compressor, options) do
+    payload =
+      Frame.new(:auth_response)
+      |> Protocol.encode_request(requested_options, options)
+      |> Frame.encode()
+
+    with :ok <- :gen_tcp.send(socket, payload),
+         {:ok, %Frame{kind: :auth_success}} <- recv_frame(socket, compressor) do
```

Comment:
Same as above. If you get `{:ok, frame}` and `frame` doesn't match the `:auth_success` frame, then it will fall down to the `else` and break. You should do `{:ok, frame}` in the `with`, then handle bad frames in a `case` (raising the "protocol violation" error if necessary).